### PR TITLE
[SPARK-39696][CORE] Fix data race in access to TaskMetrics.externalAccums 

### PR DESCRIPTION
**Pull request by eejbyfeldt**
### What changes were proposed in this pull request?
This PR fixes a data race around concurrent access to `TaskMetrics.externalAccums`. The race occurs between the `executor-heartbeater` thread and the thread executing the task. This data race is not known to cause issues on 2.12 but in 2.13 ~due this change https://github.com/scala/scala/pull/9258~ (@LuciferYang bisected this to first cause failures in scala 2.13.7 one possible reason could be https://github.com/scala/scala/pull/9786) leads to an uncaught exception in the `executor-heartbeater` thread, which means that the executor will eventually be terminated due to missing hearbeats.

This fix of using of using `CopyOnWriteArrayList` is cherry picked from https://github.com/apache/spark/pull/37206 where is was suggested as a fix by @LuciferYang since `TaskMetrics.externalAccums` is also accessed from outside the class `TaskMetrics`. The old PR was closed because at that point there was no clear understanding of the race condition. @JoshRosen commented here https://github.com/apache/spark/pull/37206#issuecomment-1189930626 saying that there should be no such race based on because all accumulators should be deserialized as part of task deserialization here: https://github.com/apache/spark/blob/0cc96f76d8a4858aee09e1fa32658da3ae76d384/core/src/main/scala/org/apache/spark/executor/Executor.scala#L507-L508 and therefore no writes should occur while the hearbeat thread will read the accumulators. But my understanding is that is incorrect as accumulators will also be deserialized as part of the taskBinary here: https://github.com/apache/spark/blob/169f828b1efe10d7f21e4b71a77f68cdd1d706d6/core/src/main/scala/org/apache/spark/scheduler/ShuffleMapTask.scala#L87-L88 which will happen while the heartbeater thread is potentially reading the accumulators. This can both due to user code using accumulators (see the new test case) but also when using the Dataframe/Dataset API as  sql metrics will also be `externalAccums`. One way metrics will be sent as part of the taskBinary is when the dep is a `ShuffleDependency`: https://github.com/apache/spark/blob/fbbcf9434ac070dd4ced4fb9efe32899c6db12a9/core/src/main/scala/org/apache/spark/Dependency.scala#L85 with a ShuffleWriteProcessor that comes from https://github.com/apache/spark/blob/fbbcf9434ac070dd4ced4fb9efe32899c6db12a9/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala#L411-L422 

### Why are the changes needed?
The current code has a data race.

### Does this PR introduce _any_ user-facing change?
It will fix an uncaught exception in the `executor-hearbeater` thread when using scala 2.13.

### How was this patch tested?
This patch adds a new test case, that before the fix was applied consistently produces the uncaught exception in the heartbeater thread when using scala 2.13.

### CODE REVIEW

**Comment by eejbyfeldt** [comment]
Ping @LuciferYang @JoshRosen that previously discussed this issue in https://github.com/apache/spark/pull/37206 let me know if would be a better approach to reopen the old PR rather then creating a new one. And/or if my description of the data race does not sounds accurate to you.

**Comment by eejbyfeldt** [comment]
> I found that before Scala 2.13.6(include) seems no this issue and the new test will failed after 2.13.7.
> 
> @eejbyfeldt I am not sure if this is caused by change of [scala/scala#9258](https://github.com/scala/scala/pull/9258), as it has been added to Scala 2.13.4.


Thanks for looking in to that closer. Should probably formulated myself more clearly that was only a guess on my part and not something I had verified. But now that you narrowed it down to 2.13.7 another possible candidate change could be https://github.com/scala/scala/pull/9786 that changed how that mutations are tracked in ArrayBuffer.

**Comment by mridulm** [comment]
This makes sense.
The TaskRunner is visible to heartbeater since it gets added to `runningTasks` before the task binary is deserialized.
During `TaskRunner.run`, we are registering the accumulators as part of the deserialization - which can result in race with the heartbearter reading the metrics.

So this is essentially a bug waiting to happen for two reasons:

* `externalAccums` is getting used in an MT-unsafe way by the task thread and the reporter thread.
* This is broken in 2.12 as well - we were just not aware of it.
  * There always was potential visibility issues in heartbeat for external accumulators.

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 1]
```diff
@@ -262,13 +263,16 @@ class TaskMetrics private[spark] () extends Serializable {
   /**
    * External accumulators registered with this task.
    */
-  @transient private[spark] lazy val externalAccums = new ArrayBuffer[AccumulatorV2[_, _]]
+  @transient private[spark] lazy val _externalAccums = new CopyOnWriteArrayList[AccumulatorV2[_, _]]
+
+  private[spark] def externalAccums() = _externalAccums.asScala
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala:268]: nit: drop the `()`

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 2]
```diff
@@ -262,13 +263,16 @@ class TaskMetrics private[spark] () extends Serializable {
...
 
   private[spark] def registerAccumulator(a: AccumulatorV2[_, _]): Unit = {
-    externalAccums += a
+    _externalAccums.add(a)
   }
 
-  private[spark] def accumulators(): Seq[AccumulatorV2[_, _]] = internalAccums ++ externalAccums
+  private[spark] def accumulators(): Seq[AccumulatorV2[_, _]] =
+    internalAccums ++ _externalAccums.asScala
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala:275]: nit: revert ?

**Review by mridulm — COMMENTED** [review]
Just a couple of minor nits, nice job chasing this down @eejbyfeldt !

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala (hunk 2 above)
Comment (eejbyfeldt) [reply core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: done

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala (hunk 1 above)
Comment (eejbyfeldt) [reply core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: done

**Review by mridulm — APPROVED** [review]
+CC @JoshRosen, @LuciferYang

**Comment by LuciferYang** [comment]
@eejbyfeldt `unused-imports` check failed, please fix it

```
[error] /home/runner/work/spark/spark/core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:48:41: Unused import
[error] import org.apache.spark.internal.config.Network
[error]                                         ^
[error] one error found
```

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 3]
```diff
@@ -206,7 +208,6 @@ class TaskMetrics private[spark] () extends Serializable {
   // Only used for test
   private[spark] val testAccum = sys.props.get(IS_TESTING.key).map(_ => new LongAccumulator)
 
-
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala:209]: please revert this change

**Comment by LuciferYang** [comment]
> I found that before Scala 2.13.6(include) seems no this issue and the new test will failed after 2.13.7.
> 
> @eejbyfeldt I am not sure if this is caused by change of [scala/scala#9258](https://github.com/scala/scala/pull/9258), as it has been added to Scala 2.13.4.
> 
> also cc @srowen @mridulm and @xinrong-meng

also cc @dongjoon-hyun due to Spark 3.2.x also use Scala 2.13.8 and maybe Spark 3.2.4 should include this one

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 4]
```diff
@@ -270,6 +272,28 @@ class ExecutorSuite extends SparkFunSuite
...
+    val conf = new SparkConf().setMaster("local").setAppName("executor suite test")
+    conf.set(EXECUTOR_HEARTBEAT_INTERVAL.key, "1ms")
+    sc = new SparkContext(conf)
+
+    val accums = (1 to 10).map(i => sc.longAccumulator(s"mapperRunAccumulator${i}"))
+    val input = sc.parallelize(1 to 10, 10)
+    val baseRdd = input.map(i => (i, i))
+    var testRdd = baseRdd
+    (0 to 10).foreach( i =>
+        testRdd = testRdd.map(x => {accums.map(_.add(1)); (x._1 * i, x._2)}).reduceByKey(_ + _)
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:285]: 1. can be`accums.foreach(_.add(1))`
2. should add spaces after `{` and before `}`, like

```
{ accums.map(_.add(1)); (x._1 * i, x._2) }
```

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 5]
```diff
@@ -270,6 +272,28 @@ class ExecutorSuite extends SparkFunSuite
...
+    )
+
+    val logAppender = new LogAppender(s"heartbeat thread shuold not die")
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:288]: 1. the string interpolator is redundant
2. `shuold` -> `should` ?

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 6]
```diff
@@ -270,6 +272,28 @@ class ExecutorSuite extends SparkFunSuite
     heartbeatZeroAccumulatorUpdateTest(false)
   }
 
+  test("SPARK-39696: Using accumulators should not cause heartbeat to fail") {
+    val conf = new SparkConf().setMaster("local").setAppName("executor suite test")
+    conf.set(EXECUTOR_HEARTBEAT_INTERVAL.key, "1ms")
+    sc = new SparkContext(conf)
+
+    val accums = (1 to 10).map(i => sc.longAccumulator(s"mapperRunAccumulator${i}"))
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:280]: The enclosing block is redundant

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 7]
```diff
@@ -270,6 +272,28 @@ class ExecutorSuite extends SparkFunSuite
...
+    val input = sc.parallelize(1 to 10, 10)
+    val baseRdd = input.map(i => (i, i))
+    var testRdd = baseRdd
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:283]: Need to assign `baseRdd` to `testRdd`?

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 8]
```diff
@@ -270,6 +271,29 @@ class ExecutorSuite extends SparkFunSuite
...
+  test("SPARK-39696: Using accumulators should not cause heartbeat to fail") {
+    val conf = new SparkConf().setMaster("local").setAppName("executor suite test")
+    conf.set(EXECUTOR_HEARTBEAT_INTERVAL.key, "1ms")
+    sc = new SparkContext(conf)
+
+    val accums = (1 to 10).map(i => sc.longAccumulator(s"mapperRunAccumulator$i"))
+    val input = sc.parallelize(1 to 10, 10)
+    var testRdd = input.map(i => (i, i))
+    (0 to 10).foreach( i =>
+        testRdd = testRdd
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:283]: indent: 2 spaces

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 9]
```diff
@@ -270,6 +271,29 @@ class ExecutorSuite extends SparkFunSuite
...
+          .map(x => { accums.foreach(_.add(1)); (x._1 * i, x._2) })
+          .reduceByKey(_ + _)
+    )
+
+    val logAppender = new LogAppender(s"heartbeat thread should not die")
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:288]:
```suggestion
    val logAppender = new LogAppender("heartbeat thread should not die")
```

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 10]
```diff
@@ -270,6 +271,29 @@ class ExecutorSuite extends SparkFunSuite
...
+    withLogAppender(logAppender, level = Some(Level.ERROR)) {
+      val _ = testRdd.count()
+    }
+    val logs = logAppender.loggingEvents.map(_.getMessage.getFormattedMessage)
+        .filter(_.contains("Uncaught exception in thread executor-heartbeater"))
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:293]: indent: 2 spaces

**Review by LuciferYang — APPROVED** [review]
+1, LGTM , pending ci

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Comment by LuciferYang** [comment]
@HyukjinKwon branch-3.3 and branch-3.2 may also require this one, they also use Scala 2.13.8, need @eejbyfeldt  to submit an independent PRs?

**Comment by dongjoon-hyun** [comment]
Thank you, @eejbyfeldt , @mridulm , @HyukjinKwon , and @LuciferYang .
I also backported to branch-3.3.

However, we need a new PR for branch-3.2 due to the compilation error, @LuciferYang and @eejbyfeldt .
```
[error] /Users/dongjoon/APACHE/spark-merge/core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala:33:19: object logging is not a member of package org.apache
[error] import org.apache.logging.log4j._
```

**Comment by dongjoon-hyun** [comment]
To @LuciferYang  and all.

After double-checking, I found that Apache Spark 3.2.x is not affected because it's Scala 2.13.5. 

https://github.com/apache/spark/blob/7773740e4141444bf78ba75dcee9f3fade7f6e11/pom.xml#L3389

SPARK-35496 (Scala 2.13.7) landed at Apache Spark 3.3.0+. We don't need to backport this to branch-3.2.

**Comment by dongjoon-hyun** [comment]
Please let me know if this is still valid in `branch-3.2`.
> branch-3.3 and branch-3.2 may also require this one, they also use Scala 2.13.8

**Comment by LuciferYang** [comment]
@dongjoon-hyun Scala 2.13.5 does not require this fix. I apologize for providing incorrect information earlier

**Comment by dongjoon-hyun** [comment]
No problem at all. Thank you always, @LuciferYang !

**Comment by mridulm** [comment]
Thanks for checking @dongjoon-hyun and @LuciferYang !
Great to finally have this issue fixed :-)
